### PR TITLE
Linux Hotplug: Connection-callback implementation for hidraw

### PR DESCRIPTION
**Pull request by k1-801**

### CODE REVIEW

File: linux/hid.c
Comment (Youw) [inline linux/hid.c:1]: HIDAPI uses tabs for indentation
if you've noticed otherwise - thet's a "bug"

File: linux/hid.c [hunk 1]
```diff
@@ -942,26 +948,239 @@ void  HID_API_EXPORT hid_free_enumeration(struct hid_device_info *devs)
...
+    .monitor_fd = -1,
+    .next_handle = 1,
+    .hotplug_cbs = NULL,
+    .devs = NULL
+};
+
+static void hid_internal_invoke_callbacks(struct hid_device_info* info, hid_hotplug_event event)
+{
+    struct hid_hotplug_callback **current = &hid_hotplug_context.hotplug_cbs;
+    while(*current) {
```
Comment (Youw) [inline linux/hid.c:998]:
```suggestion
    while (*current) {
```
always a space after `if`/`for`/`while`/etc.

File: linux/hid.c [hunk 2]
```diff
@@ -942,26 +948,239 @@ void  HID_API_EXPORT hid_free_enumeration(struct hid_device_info *devs)
...
+    if(!strcmp(path, info->path)) {
+        return 1;
+    }
+    return 0;
+}
+
+static void* hotplug_thread(void*)
+{
+    while(hid_hotplug_context.monitor_fd > 0)
+    {
```
Comment (Youw) [inline linux/hid.c:1033]:
```suggestion
    while (hid_hotplug_context.monitor_fd > 0) {
```
consistency

**Review by Youw — COMMENTED** [review]
my comments are only about code-style for now
I really need to find time to review this (and others) thoroughly

**Comment by k1-801** [comment]
Did some adjustents to the code-style (in fact, got my IDE set up to keep track of it).
Made a few changes to the list management code, as the old version had a severe logical error in device removal handling.

File: linux/hid.c [hunk 3]
```diff
@@ -818,6 +823,82 @@ HID_API_EXPORT const char* HID_API_CALL hid_version_str(void)
...
+}
+
+static void hid_internal_hotplug_exit()
+{
+	if (!hid_hotplug_context.mutex_ready) {
+		return;
+	}
+
+	pthread_mutex_lock(&hid_hotplug_context.mutex);
+	hid_hotplug_callback** current = &hid_hotplug_context.hotplug_cbs
```
Comment (k1-801) [inline linux/hid.c:889]: I have no idea how did this wasn't fixed in the commit despite this code having actually been run and tested. Will look at what other changes didn't get commited.

**Comment by mcuee** [comment]
Not so sure how to re-run the github actions as there were build error previously.

**Comment by mcuee** [comment]
I will test this under Linux later.

**Comment by mcuee** [comment]
This seems to work fine under Ubuntu Linux 20.04.

Unplug and plug of Xiaomi USB mouse receiver

Unplug and plug of USBasp programmer (USB Composite device)

Unplug and plug of Plantronics Headset (USB composite device)

```
mcuee@UbuntuSwift3 ~/build/hid/hidapi_k1801 (connection-callback)$ ./hidtest/hidtest-hidraw 
hidapi test/example tool. Compiled with hidapi version 0.14.0, runtime version 0.14.0.
Compile-time version matches runtime version of hidapi.

Device Found
  type: 16c0 05dc
  path: /dev/hidraw2
  serial_number: 0001
  Manufacturer: www.fischl.de
  Product:      USBasp
  Release:      110
  Interface:    1
  Usage (page): 0x1 (0xff00)
  Bus type: 1 (USB)

  Report Descriptor: Unable to open device by path
Device Found
  type: 1ea7 0064
  path: /dev/hidraw0
  serial_number: 
  Manufacturer: (null)
  Product:      2.4G Mouse
  Release:      200
  Interface:    0
  Usage (page): 0x1 (0xffb5)
  Bus type: 1 (USB)

  Report Descriptor: Unable to open device by path
Device Found
  type: 1ea7 0064
  path: /dev/hidraw0
  serial_number: 
  Manufacturer: (null)
  Product:      2.4G Mouse
  Release:      200
  Interface:    0
  Usage (page): 0x2 (0x1)
  Bus type: 1 (USB)

  Report Descriptor: Unable to open device by path
Device Found
  type: 06cb cd40
  path: /dev/hidraw1
  serial_number: 
  Manufacturer: 
  Product:      SYNA7DAB:01 06CB:CD40
  Release:      0
  Interface:    -1
  Usage (page): 0x2 (0x1)
  Bus type: 3 (I2C)

  Report Descriptor: Unable to open device by path
Device Found
  type: 06cb cd40
  path: /dev/hidraw1
  serial_number: 
  Manufacturer: 
  Product:      SYNA7DAB:01 06CB:CD40
  Release:      0
  Interface:    -1
  Usage (page): 0x5 (0xd)
  Bus type: 3 (I2C)

  Report Descriptor: Unable to open device by path
Device Found
  type: 06cb cd40
  path: /dev/hidraw1
  serial_number: 
  Manufacturer: 
  Product:      SYNA7DAB:01 06CB:CD40
  Release:      0
  Interface:    -1
  Usage (page): 0xe (0xd)
  Bus type: 3 (I2C)

  Report Descriptor: Unable to open device by path
Device Found
  type: 06cb cd40
  path: /dev/hidraw1
  serial_number: 
  Manufacturer: 
  Product:      SYNA7DAB:01 06CB:CD40
  Release:      0
  Interface:    -1
  Usage (page): 0x1 (0xff00)
  Bus type: 3 (I2C)

  Report Descriptor: Unable to open device by path
Handle 1: Device was disconnected: /dev/hidraw2.
type: 16c0 05dc
  serial_number: 0001
  Manufacturer: www.fischl.de
  Product:      USBasp
  Release:      110
  Interface:    1
  Usage (page): 0x1 (0xff00)

Handle 1: New device is connected: /dev/hidraw2.
type: 16c0 05dc
  serial_number: 0001
  Manufacturer: www.fischl.de
  Product:      USBasp
  Release:      110
  Interface:    1
  Usage (page): 0x1 (0xff00)

Handle 1: Device was disconnected: /dev/hidraw0.
type: 1ea7 0064
  serial_number: 
  Manufacturer: (null)
  Product:      2.4G Mouse
  Release:      200
  Interface:    0
  Usage (page): 0x1 (0xffb5)

Handle 1: Device was disconnected: /dev/hidraw0.
type: 1ea7 0064
  serial_number: 
  Manufacturer: (null)
  Product:      2.4G Mouse
  Release:      200
  Interface:    0
  Usage (page): 0x2 (0x1)

Handle 1: New device is connected: /dev/hidraw0.
type: 1ea7 0064
  serial_number: 
  Manufacturer: (null)
  Product:      2.4G Mouse
  Release:      200
  Interface:    0
  Usage (page): 0x1 (0xffb5)

Handle 1: New device is connected: /dev/hidraw0.
type: 1ea7 0064
  serial_number: 
  Manufacturer: (null)
  Product:      2.4G Mouse
  Release:      200
  Interface:    0
  Usage (page): 0x2 (0x1)

Handle 1: Device was disconnected: /dev/hidraw2.
type: 16c0 05dc
  serial_number: 0001
  Manufacturer: www.fischl.de
  Product:      USBasp
  Release:      110
  Interface:    1
  Usage (page): 0x1 (0xff00)

Handle 1: New device is connected: /dev/hidraw2.
type: 047f c056
  serial_number: BB305534D79EBD418B6757528EAC0C19
  Manufacturer: Plantronics
  Product:      Plantronics Blackwire 3220 Series
  Release:      210
  Interface:    3
  Usage (page): 0x1 (0xc)

Handle 1: New device is connected: /dev/hidraw2.
type: 047f c056
  serial_number: BB305534D79EBD418B6757528EAC0C19
  Manufacturer: Plantronics
  Product:      Plantronics Blackwire 3220 Series
  Release:      210
  Interface:    3
  Usage (page): 0x5 (0xb)

Handle 1: New device is connected: /dev/hidraw2.
type: 047f c056
  serial_number: BB305534D79EBD418B6757528EAC0C19
  Manufacturer: Plantronics
  Product:      Plantronics Blackwire 3220 Series
  Release:      210
  Interface:    3
  Usage (page): 0x3 (0xffa0)

Handle 1: Device was disconnected: /dev/hidraw2.
type: 047f c056
  serial_number: BB305534D79EBD418B6757528EAC0C19
  Manufacturer: Plantronics
  Product:      Plantronics Blackwire 3220 Series
  Release:      210
  Interface:    3
  Usage (page): 0x1 (0xc)

Handle 1: Device was disconnected: /dev/hidraw2.
type: 047f c056
  serial_number: BB305534D79EBD418B6757528EAC0C19
  Manufacturer: Plantronics
  Product:      Plantronics Blackwire 3220 Series
  Release:      210
  Interface:    3
  Usage (page): 0x5 (0xb)

Handle 1: Device was disconnected: /dev/hidraw2.
type: 047f c056
  serial_number: BB305534D79EBD418B6757528EAC0C19
  Manufacturer: Plantronics
  Product:      Plantronics Blackwire 3220 Series
  Release:      210
  Interface:    3
  Usage (page): 0x3 (0xffa0)

^C

```

**Comment by k1-801** [comment]
Noticed a very minor issue with this PR - the code lacks a `pthread_join` in `hotlplug_cleanup`. Might cause some weird behaviour in a very narrow case when all callbacks are deleted and then a new one is installed immediately and the background thread didn't have enough time to stop. Going to fix in a moment.

**Comment by k1-801** [comment]
Fixed the cleanup procedure. Doesn't affect how the events themselves work.

**Comment by mcuee** [comment]
Last git commit is still working fine. Thanks. 

Tested under Ubuntu Linux 20.04.

Unplug and plug of Platronics USB headset
Unplug and plug of a USB mouse receiver
Unplug and plug of another USB mouse receiver

<details>

```
mcuee@UbuntuSwift3 ~/build/hid/hidapi_hotplug_hidraw (connection-callback-udev)$ ./hidtest/hidtest-hidraw 
hidapi test/example tool. Compiled with hidapi version 0.14.0, runtime version 0.14.0.
Compile-time version matches runtime version of hidapi.

Device Found
  type: 1ea7 0066
  path: /dev/hidraw1
  serial_number: 
  Manufacturer: (null)
  Product:      2.4G Mouse
  Release:      200
  Interface:    0
  Usage (page): 0x6 (0x1)
  Bus type: 1 (USB)

  Report Descriptor: Unable to open device by path
Device Found
  type: 1ea7 0066
  path: /dev/hidraw2
  serial_number: 
  Manufacturer: (null)
  Product:      2.4G Mouse
  Release:      200
  Interface:    1
  Usage (page): 0x1 (0xffb5)
  Bus type: 1 (USB)

  Report Descriptor: Unable to open device by path
Device Found
  type: 1ea7 0066
  path: /dev/hidraw2
  serial_number: 
  Manufacturer: (null)
  Product:      2.4G Mouse
  Release:      200
  Interface:    1
  Usage (page): 0x2 (0x1)
  Bus type: 1 (USB)

  Report Descriptor: Unable to open device by path
Device Found
  type: 1ea7 0066
  path: /dev/hidraw2
  serial_number: 
  Manufacturer: (null)
  Product:      2.4G Mouse
  Release:      200
  Interface:    1
  Usage (page): 0x1 (0xc)
  Bus type: 1 (USB)

  Report Descriptor: Unable to open device by path
Device Found
  type: 1ea7 0066
  path: /dev/hidraw2
  serial_number: 
  Manufacturer: (null)
  Product:      2.4G Mouse
  Release:      200
  Interface:    1
  Usage (page): 0x80 (0x1)
  Bus type: 1 (USB)

  Report Descriptor: Unable to open device by path
Device Found
  type: 1ea7 0066
  path: /dev/hidraw2
  serial_number: 
  Manufacturer: (null)
  Product:      2.4G Mouse
  Release:      200
  Interface:    1
  Usage (page): 0x88 (0xffbc)
  Bus type: 1 (USB)

  Report Descriptor: Unable to open device by path
Device Found
  type: 047f c056
  path: /dev/hidraw4
  serial_number: D1CEC32927974D5F9BD6B2AEBF2EA8E3
  Manufacturer: Plantronics
  Product:      Plantronics Blackwire 3220 Series
  Release:      210
  Interface:    3
  Usage (page): 0x1 (0xc)
  Bus type: 1 (USB)

  Report Descriptor: Unable to open device by path
Device Found
  type: 047f c056
  path: /dev/hidraw4
  serial_number: D1CEC32927974D5F9BD6B2AEBF2EA8E3
  Manufacturer: Plantronics
  Product:      Plantronics Blackwire 3220 Series
  Release:      210
  Interface:    3
  Usage (page): 0x5 (0xb)
  Bus type: 1 (USB)

  Report Descriptor: Unable to open device by path
Device Found
  type: 047f c056
  path: /dev/hidraw4
  serial_number: D1CEC32927974D5F9BD6B2AEBF2EA8E3
  Manufacturer: Plantronics
  Product:      Plantronics Blackwire 3220 Series
  Release:      210
  Interface:    3
  Usage (page): 0x3 (0xffa0)
  Bus type: 1 (USB)

  Report Descriptor: Unable to open device by path
Device Found
  type: 1ea7 0064
  path: /dev/hidraw3
  serial_number: 
  Manufacturer: (null)
  Product:      2.4G Mouse
  Release:      200
  Interface:    0
  Usage (page): 0x1 (0xffb5)
  Bus type: 1 (USB)

  Report Descriptor: Unable to open device by path
Device Found
  type: 1ea7 0064
  path: /dev/hidraw3
  serial_number: 
  Manufacturer: (null)
  Product:      2.4G Mouse
  Release:      200
  Interface:    0
  Usage (page): 0x2 (0x1)
  Bus type: 1 (USB)

  Report Descriptor: Unable to open device by path
Device Found
  type: 06cb cd40
  path: /dev/hidraw0
  serial_number: 
  Manufacturer: 
  Product:      SYNA7DAB:01 06CB:CD40
  Release:      0
  Interface:    -1
  Usage (page): 0x2 (0x1)
  Bus type: 3 (I2C)

  Report Descriptor: Unable to open device by path
Device Found
  type: 06cb cd40
  path: /dev/hidraw0
  serial_number: 
  Manufacturer: 
  Product:      SYNA7DAB:01 06CB:CD40
  Release:      0
  Interface:    -1
  Usage (page): 0x5 (0xd)
  Bus type: 3 (I2C)

  Report Descriptor: Unable to open device by path
Device Found
  type: 06cb cd40
  path: /dev/hidraw0
  serial_number: 
  Manufacturer: 
  Product:      SYNA7DAB:01 06CB:CD40
  Release:      0
  Interface:    -1
  Usage (page): 0xe (0xd)
  Bus type: 3 (I2C)

  Report Descriptor: Unable to open device by path
Device Found
  type: 06cb cd40
  path: /dev/hidraw0
  serial_number: 
  Manufacturer: 
  Product:      SYNA7DAB:01 06CB:CD40
  Release:      0
  Interface:    -1
  Usage (page): 0x1 (0xff00)
  Bus type: 3 (I2C)

  Report Descriptor: Unable to open device by path
Handle 1: Device was disconnected: /dev/hidraw4.
type: 047f c056
  serial_number: D1CEC32927974D5F9BD6B2AEBF2EA8E3
  Manufacturer: Plantronics
  Product:      Plantronics Blackwire 3220 Series
  Release:      210
  Interface:    3
  Usage (page): 0x1 (0xc)

Handle 1: Device was disconnected: /dev/hidraw4.
type: 047f c056
  serial_number: D1CEC32927974D5F9BD6B2AEBF2EA8E3
  Manufacturer: Plantronics
  Product:      Plantronics Blackwire 3220 Series
  Release:      210
  Interface:    3
  Usage (page): 0x5 (0xb)

Handle 1: Device was disconnected: /dev/hidraw4.
type: 047f c056
  serial_number: D1CEC32927974D5F9BD6B2AEBF2EA8E3
  Manufacturer: Plantronics
  Product:      Plantronics Blackwire 3220 Series
  Release:      210
  Interface:    3
  Usage (page): 0x3 (0xffa0)

Handle 1: New device is connected: /dev/hidraw4.
type: 047f c056
  serial_number: D1CEC32927974D5F9BD6B2AEBF2EA8E3
  Manufacturer: Plantronics
  Product:      Plantronics Blackwire 3220 Series
  Release:      210
  Interface:    3
  Usage (page): 0x1 (0xc)

Handle 1: New device is connected: /dev/hidraw4.
type: 047f c056
  serial_number: D1CEC32927974D5F9BD6B2AEBF2EA8E3
  Manufacturer: Plantronics
  Product:      Plantronics Blackwire 3220 Series
  Release:      210
  Interface:    3
  Usage (page): 0x5 (0xb)

Handle 1: New device is connected: /dev/hidraw4.
type: 047f c056
  serial_number: D1CEC32927974D5F9BD6B2AEBF2EA8E3
  Manufacturer: Plantronics
  Product:      Plantronics Blackwire 3220 Series
  Release:      210
  Interface:    3
  Usage (page): 0x3 (0xffa0)

Handle 1: Device was disconnected: /dev/hidraw3.
type: 1ea7 0064
  serial_number: 
  Manufacturer: (null)
  Product:      2.4G Mouse
  Release:      200
  Interface:    0
  Usage (page): 0x1 (0xffb5)

Handle 1: Device was disconnected: /dev/hidraw3.
type: 1ea7 0064
  serial_number: 
  Manufacturer: (null)
  Product:      2.4G Mouse
  Release:      200
  Interface:    0
  Usage (page): 0x2 (0x1)

Handle 1: New device is connected: /dev/hidraw3.
type: 1ea7 0064
  serial_number: 
  Manufacturer: (null)
  Product:      2.4G Mouse
  Release:      200
  Interface:    0
  Usage (page): 0x1 (0xffb5)

Handle 1: New device is connected: /dev/hidraw3.
type: 1ea7 0064
  serial_number: 
  Manufacturer: (null)
  Product:      2.4G Mouse
  Release:      200
  Interface:    0
  Usage (page): 0x2 (0x1)

Handle 1: Device was disconnected: /dev/hidraw1.
type: 1ea7 0066
  serial_number: 
  Manufacturer: (null)
  Product:      2.4G Mouse
  Release:      200
  Interface:    0
  Usage (page): 0x6 (0x1)

Handle 1: Device was disconnected: /dev/hidraw2.
type: 1ea7 0066
  serial_number: 
  Manufacturer: (null)
  Product:      2.4G Mouse
  Release:      200
  Interface:    1
  Usage (page): 0x1 (0xffb5)

Handle 1: Device was disconnected: /dev/hidraw2.
type: 1ea7 0066
  serial_number: 
  Manufacturer: (null)
  Product:      2.4G Mouse
  Release:      200
  Interface:    1
  Usage (page): 0x2 (0x1)

Handle 1: Device was disconnected: /dev/hidraw2.
type: 1ea7 0066
  serial_number: 
  Manufacturer: (null)
  Product:      2.4G Mouse
  Release:      200
  Interface:    1
  Usage (page): 0x1 (0xc)

Handle 1: Device was disconnected: /dev/hidraw2.
type: 1ea7 0066
  serial_number: 
  Manufacturer: (null)
  Product:      2.4G Mouse
  Release:      200
  Interface:    1
  Usage (page): 0x80 (0x1)

Handle 1: Device was disconnected: /dev/hidraw2.
type: 1ea7 0066
  serial_number: 
  Manufacturer: (null)
  Product:      2.4G Mouse
  Release:      200
  Interface:    1
  Usage (page): 0x88 (0xffbc)

Handle 1: New device is connected: /dev/hidraw2.
type: 1ea7 0066
  serial_number: 
  Manufacturer: (null)
  Product:      2.4G Mouse
  Release:      200
  Interface:    1
  Usage (page): 0x1 (0xffb5)

Handle 1: New device is connected: /dev/hidraw2.
type: 1ea7 0066
  serial_number: 
  Manufacturer: (null)
  Product:      2.4G Mouse
  Release:      200
  Interface:    1
  Usage (page): 0x2 (0x1)

Handle 1: New device is connected: /dev/hidraw2.
type: 1ea7 0066
  serial_number: 
  Manufacturer: (null)
  Product:      2.4G Mouse
  Release:      200
  Interface:    1
  Usage (page): 0x1 (0xc)

Handle 1: New device is connected: /dev/hidraw2.
type: 1ea7 0066
  serial_number: 
  Manufacturer: (null)
  Product:      2.4G Mouse
  Release:      200
  Interface:    1
  Usage (page): 0x80 (0x1)

Handle 1: New device is connected: /dev/hidraw2.
type: 1ea7 0066
  serial_number: 
  Manufacturer: (null)
  Product:      2.4G Mouse
  Release:      200
  Interface:    1
  Usage (page): 0x88 (0xffbc)

Handle 1: New device is connected: /dev/hidraw1.
type: 1ea7 0066
  serial_number: 
  Manufacturer: (null)
  Product:      2.4G Mouse
  Release:      200
  Interface:    0
  Usage (page): 0x6 (0x1)

Handle 1: New device is connected: /dev/hidraw5.
type: 16c0 05dc
  serial_number: 0001
  Manufacturer: www.fischl.de
  Product:      USBasp
  Release:      110
  Interface:    1
  Usage (page): 0x1 (0xff00)

Handle 1: Device was disconnected: /dev/hidraw5.
type: 16c0 05dc
  serial_number: 0001
  Manufacturer: www.fischl.de
  Product:      USBasp
  Release:      110
  Interface:    1
  Usage (page): 0x1 (0xff00)

^C
```

</details>

**Review by Youw — APPROVED** [review]
Same deal as with #645 - getting it merged into a feature branch now.

**Pull request merged**